common/networkselectionmodel.cpp

```diff
@@ -43,87 +49,166 @@ static void writeSelection(Message *msg, const QItemSelection &selection)
   }
 }
 
+static bool isMainWindowSubclassAcceptor(const QVariant &v)
+{
+  const QObject *object = v.value<QObject*>();
+  return object && object->inherits("QMainWindow");
+}
+
 NetworkSelectionModel::NetworkSelectionModel(const QString &objectName, QAbstractItemModel* model, QObject* parent):
   QItemSelectionModel(model, parent),
   m_objectName(objectName),
   m_myAddress(Protocol::InvalidObjectAddress),
   m_handlingRemoteMessage(false)
 {
+  setObjectName(m_objectName + QLatin1String("Network"));
   connect(this, SIGNAL(currentChanged(QModelIndex,QModelIndex)), this, SLOT(slotCurrentChanged(QModelIndex,QModelIndex)));
   connect(this, SIGNAL(currentColumnChanged(QModelIndex,QModelIndex)), this, SLOT(slotCurrentColumnChanged(QModelIndex,QModelIndex)));
   connect(this, SIGNAL(currentRowChanged(QModelIndex,QModelIndex)), this, SLOT(slotCurrentRowChanged(QModelIndex,QModelIndex)));
   connect(this, SIGNAL(selectionChanged(QItemSelection,QItemSelection)), this, SLOT(slotSelectionChanged(QItemSelection,QItemSelection)));
-
-  Q_ASSERT(model);
-  connect(model, SIGNAL(modelAboutToBeReset()), this, SLOT(clearPendingSelection()));
-  connect(model, SIGNAL(rowsInserted(QModelIndex,int,int)), this, SLOT(applyPendingSelection()));
 }
 
 NetworkSelectionModel::~NetworkSelectionModel()
 {
 }
 
-Protocol::ItemSelection GammaRay::NetworkSelectionModel::readSelection(const GammaRay::Message& msg)
+void NetworkSelectionModel::requestSelection()
+{
+  if (m_handlingRemoteMessage || !Endpoint::isConnected() || m_myAddress == Protocol::InvalidObjectAddress) {
+    return;
+  }
+  Message msg(m_myAddress, Protocol::SelectionModelStateRequest);
+  Endpoint::send(msg);
+}
+
+void NetworkSelectionModel::sendSelection()
 {
-    Protocol::ItemSelection selection;
-    qint32 size = 0;
-    msg.payload() >> size;
-    selection.reserve(size);
-
-    for (int i = 0; i < size; ++i) {
-        Protocol::ItemSelectionRange range;
-        msg.payload() >> range.topLeft >> range.bottomRight;
-        selection.push_back(range);
+  if (!Endpoint::isConnected() || m_myAddress == Protocol::InvalidObjectAddress) {
+    return;
+  }
+
+  clearPendingSelection();
+
+  const QItemSelectionModel::SelectionFlags flags = QItemSelectionModel::ClearAndSelect |
+      QItemSelectionModel::Rows | QItemSelectionModel::Current;
+
+  if (!hasSelection()) {
+    // Only match tools selection: com.kdab.GammaRay.XXX.selection
```

Comment:
Right, I will find a way to put this back to the corresponding plugin for specific case.
Though selecting index(0,0) can stay there i guess (generic) ?